server/graphql/queries.js

```diff
@@ -15,14 +15,36 @@ export default () => [Query, Station, Stop]
 
 export const resolvers = {
   Query: {
-    stations (_, { search = '', from, length }) {
-      return db().query(aql`
+    stations (_, { search = '', line = '', from, length }) {
+      return db().query(
+        aql`
+          let route_ids = (
+              for route in routes
+              return route.route_id
+          )
+
+          let trip_ids_by_route_id = (
+            for trip in trips
+            filter (trip.route_id == ${line}) || (${line} == '' && trip.route_id in route_ids)
+            return trip.trip_id)
+
+          let stop_ids_by_trip = (
+            for stop_time in stop_times
+            filter stop_time.trip_id in trip_ids_by_route_id
+            return stop_time.stop_id)
+
+          let parent_stations = (
+            for stop in stops
+            filter stop.stop_id in stop_ids_by_trip
+            return stop.parent_station)
+
           for stop in (${search} ? fulltext(stops, "stop_name", concat("prefix:", ${search})) : stops)
-          filter stop.location_type == 1
+          filter stop.location_type == 1 && stop.stop_id in parent_stations
```

Comment:
Est-ce que ça, ça marcherait ?
1. Ligne 21, supprimer `let route_ids = (...)`
2. Ligne 28, ne garder que `trip.route_id == ${line}`
3. Ligne 42, remplacer par `filter stop.location_type == 1 && (${line} == '' || stop.stop_id in parent_stations)`

Comme ça si `line` n'est pas renseignée,  `trip_ids_by_route_id` est vide et le calcul de `stop_ids_by_trip` et `parent_stations` est quasi-instantané, non ?